### ISSUE TITLE
PECL

### ISSUE THREAD

**Issue by rtheunissen** [php-ds/extension#6]
I would like to manage releases and installation through PECL. I'm not sure what is involved exactly, but will take a look at it when the docs at #5 are completed.

@krakjoe might be a good thing to chat about.

**Comment by krakjoe** [comment]
I have created the package in PECL ...

  - [Request a pecl account](https://pecl.php.net/account-request.php)
 -  I will set you as lead on package.
  - Add package.xml for PECL (I can do that)
  - Profit.

**Comment by rtheunissen** [comment]
:money_with_wings:

**Comment by rtheunissen** [comment]
I've submitted a request, will report on progress.

**Comment by krakjoe** [comment]
Set as lead on package.

You want to have a go at package.xml, or want me to commit initial one ?

**Comment by rtheunissen** [comment]
I'll make an attempt, would be good to become familiar.

**Comment by krakjoe** [comment]
Good (read: correct) call ;)

**Comment by rtheunissen** [comment]
:sparkles:

**Comment by rtheunissen** [comment]
Added draft package.xml

Q:
1. Should we be using the PHP license?
2. Why do we have to list all the files?
3. Is an "alpha" release a valid release type?

PS: Files will change during refactor.

**Comment by krakjoe** [comment]
> Should we be using the PHP license?

While the project is outside of php-src, the MIT license is fine, causes no problems with dist and packagist process.

When the project is part of php-src, it will be a good idea to swap over to the PHP license.

If that happens early, there is no loss ...

> Why do we have to list all the files?

The ```package.xml``` file is used to create release tarballs, the files you include in there are included in the release tarball (composer.json can be left out, so can .travis.yml etc).

> Is an "alpha" release a valid release type?

As far as I know, although I've never used it ...

> PS: Files will change during refactor.

That's fine, this is only for practice ;)

**Comment by rtheunissen** [comment]
> The required minimum version of php in package.xml should be adjusted past the inclusion of the resolution for #16

Can't find a clear closed bug report for it..

**Comment by krakjoe** [comment]
Ping @nikic, can you dig up that commit for us please Nikita ?

**Comment by rtheunissen** [comment]
Linked in #16 

The bug report notes that `allow_null` is true fixes it. I noticed that the arginfo entry was the only case where `allow_null` was false. I've updated the arginfo to enable `allow_null`, so our version is not affected by the bug.

What's the best way to determine min required version? Build and test?

**Comment by listepo** [comment]
@rtheunissen what do you think about the .deb and rpm packages?

**Comment by shouze** [comment]
@rtheunissen any chance to push the 1st release on pecl soon? We're using ds since about 2 months on some project and (except maybe some leaks, we didn't check well) we didn't encounter any problem.

**Comment by rtheunissen** [comment]
@shouze I'm working on this every day as we speak. My focus at the moment is on source refactoring, leaving only documentation left. I'm not opposed to an early PECL release without proper documentation, but it's definitely something I hope to get out in the coming weeks.

**Comment by shouze** [comment]
I understand, maybe I can help on documentation? Not very familiar with php documentation but no pb to GET familiar with quickly.

**Comment by rtheunissen** [comment]
Most of the documentation is already written under `php/include`, it's just not in the proper docbook XML format - something which I'm not personally familiar with either. It's not a requirement for PECL (as far as I know), so I'm open to a pre-stable release without complete documentation. This would also be a good opportunity to "practice" the release schedule. 

I don't want to release a stable 1.0.0 without fully integrated XML documentation.

**Comment by shouze** [comment]
Ok I'm pretty familiar with docbook I've written some 10 years ago, just have to remember it, will give a try with existing documentation.

**Comment by rtheunissen** [comment]
Actually it used to be under `php/include`, it's now just under the [polyfill](https://github.com/php-ds/polyfill) repo.

**Comment by rtheunissen** [comment]
@shouze have you started on the documentation at all? I'm about to, just wanted to ask before I potentially redo some of what you've already done. 

Will be working on it at https://github.com/php-ds/extension/tree/docbook

**Comment by shouze** [comment]
@rtheunissen not started yet.

**Comment by rtheunissen** [comment]
@shouze I've created https://github.com/php-ds/docs, will commit the boilerplate tomorrow.

**Comment by rtheunissen** [comment]
@shouze See https://github.com/php-ds/docs

**Comment by rtheunissen** [comment]
Documentation is coming along nicely. Currently working on `Deque`, which I'll copy over to `Sequence` and `Vector` with a few changes. 

That leaves `Map`, `Stack`, `Queue`, `PriorityQueue` and `Set`. 

@shouze, welcome to help out here. I'll try to get it all done by the end of the weekend anyway, but any help here would fast-track the PECL release.

**Comment by shouze** [comment]
@rtheunissen great, I've forked https://github.com/php-ds/docs and I take on Map at least this week ;)

**Comment by rtheunissen** [comment]
@shouze awesome! Let me know if anything is unclear.

**Comment by rtheunissen** [comment]
@shouze I've done the docs for `Map` now, leaving only a few small classes and introductions that I'll cover tomorrow. If you have some time and would like to contribute, feel free to pull and render, and create issues for things that are either incorrect, misleading, or confusing. 😄

**Comment by rtheunissen** [comment]
All docs are now complete. We're ready for PECL. ✨

**Comment by shouze** [comment]
@rtheunissen waw, what fast you are! ;)

I've fetched & builded locally, I've just noticed something strange ATM:

Despite you have that in docs/versions.xml:
```xml
<function name='ds\collection' from='PECL ds &gt;= 1.0.0'/>
```

Once I build the page looks like:

<img width="392" alt="capture d ecran 2016-07-27 a 21 20 04" src="https://cloud.githubusercontent.com/assets/54712/17189144/87a50236-5440-11e6-953f-56fc74f41823.png">

But it works perfectly for
```xml
<function name='ds\sequence::__construct' from='PECL ds &gt;= 1.0.0'/>
```

**Comment by rtheunissen** [comment]
@shouze I noticed that too, and I have no idea why the version information doesn't show up for the classes. There's also https://github.com/php-ds/docs/issues/4 which I'm still trying to debug. I'll open an issue there - let's move discussion over to the docs repo for now.

**Comment by rtheunissen** [comment]
@krakjoe mentioned in room 11 but will copy here for the record. Everything has been fixed, the build on PECL is 1.0.1, and I'm waiting to see if the Windows build completes.

Q: Seeing as 1.0.0 is broken and 1.0.1 wasn't released as "stable", is it possible to remove those releases or start over? Or should we just release 1.0.2 as stable and go from there?

**Comment by rtheunissen** [comment]
Also, do you manually edit package.xml or is there an interface / generator for it? Noticed APCu has a changelog in the package.xml as well. Manual editing or generated?

**Comment by rtheunissen** [comment]
I'm ready to release as stable and submit the documentation, on standby.

**Comment by rtheunissen** [comment]
The only things left to discuss are:

- The possibility of purging the current 1.0.0 and 1.0.1 devel releases in favour of a 1.0.0 stable release. Could nuke the entire package and create a new one with the same name. If that's not a possibility (or a bad idea), I'm in favour of a 1.0.3 stable release.
- When would we use `pecl svntag`
- When would we use `pecl sign`

**Comment by rtheunissen** [comment]
Released 1.0.2 as stable. Can now install via `pecl install ds` or download DLL's.

**Comment by shouze** [comment]
@rtheunissen great work, thank you so much!
